snuba/query/processors/uuid_column_processor.py

```diff
@@ -26,6 +27,141 @@
+            except (AssertionError, ValueError):
+                raise UUIDTranslationError("Not a valid UUID string")
+
+        def process_condition(exp: Expression) -> Expression:
+            match = self.__condition_matcher.match(exp)
+
+            if match:
+                try:
+                    return FunctionCall(
+                        exp.alias,
+                        match.string("operator"),
+                        (
+                            match.expression("col"),
+                            translate_uuid_literal(match.expression("literal")),
+                        ),
+                    )
+                except UUIDTranslationError:
+                    # We probably won't get to this point since we should have validations earlier
+                    # in the pipeline, but just in case return 0 otherwise the query is going to
+                    # cause a type mismatch on ClickHouse
+                    return Literal(exp.alias, 0)
+
+            in_condition_match = self.__in_condition_matcher.match(exp)
+
+            if in_condition_match:
+                try:
+                    tuple_func = in_condition_match.expression("tuple")
+                    assert isinstance(tuple_func, FunctionCall)
+                    new_tuple_func = FunctionCall(
+                        tuple_func.alias,
+                        tuple_func.function_name,
+                        parameters=tuple(
+                            [
+                                translate_uuid_literal(lit)
+                                for lit in tuple_func.parameters
+                            ]
+                        ),
+                    )
+                    return FunctionCall(
+                        exp.alias,
+                        in_condition_match.string("operator"),
+                        (in_condition_match.expression("col"), new_tuple_func,),
+                    )
+                except UUIDTranslationError:
+                    return Literal(exp.alias, 0)
+
+            return exp
+
+        def process_all(exp: Expression) -> Expression:
+            if isinstance(exp, Column):
+                if exp.column_name in self.__uuid_columns:
+                    return FunctionCall(
+                        exp.alias,
+                        "replaceAll",
+                        (
+                            FunctionCall(
+                                None,
+                                "toString",
+                                (Column(None, None, exp.column_name),),
+                            ),
+                            Literal(None, "-"),
+                            Literal(None, ""),
+                        ),
+                    )
+
+            return exp
+
+        query.transform_expressions(process_all)
+
+        condition = query.get_condition()
+        if condition:
+            query.set_ast_condition(condition.transform(process_condition))
```

Comment:
It would be nice to do this in one pass.
We know that all the expression to optimize are in the condition. And that All those expressions that include the UUID that were not optimized need to have `replaceAll` applied. 
If we made the change attribute by attribute on the query we could apply a simple rule to all the clauses, and apply a different one to the condition (that would do the optimization). 
This would make the patterns simpler because we would nbever have to match an expresison that is already turned into a `replaceAll`. Though it would couple this with the list of clauses in the query (which is unlikely to change).

What do you think ?

Reply:
This was my original goal and I spent a really long time trying to find a non painful way to achieve this. I think having special handling for every single query clause (essentially implementing our own special version of `query.transform_expressions()` inside the processor is going to be very long winded and I wanted to avoid it. Ideally I think the query class could provide some way to do this because it may be useful in other processors. I'm not sure how this should work though. Perhaps either by providing a way to omit some parts of the query in `transform_expressions`, or by telling the transform function what part of the query it's in while it's traversing the expressions. But I haven't thought too much about this, and I'm not entirely sure if it's a reasonable idea yet.